consensus/src/main/java/org/ethereum/beacon/consensus/util/CachingBeaconChainSpec.java

```diff
@@ -2,57 +2,48 @@
 
 import java.util.List;
 import java.util.Map;
-import java.util.Objects;
 import java.util.concurrent.ConcurrentHashMap;
-import java.util.function.Function;
-import javax.annotation.Nonnull;
-import org.ethereum.beacon.consensus.SpecHelpers;
-import org.ethereum.beacon.consensus.hasher.ObjectHasher;
-import org.ethereum.beacon.consensus.hasher.SSZObjectHasher;
+import org.ethereum.beacon.consensus.BeaconChainSpec;
+import org.ethereum.beacon.consensus.DelegatingBeaconChainSpec;
 import org.ethereum.beacon.core.BeaconState;
-import org.ethereum.beacon.core.spec.SpecConstants;
 import org.ethereum.beacon.core.types.BLSPubkey;
 import org.ethereum.beacon.core.types.ValidatorIndex;
-import org.ethereum.beacon.crypto.Hashes;
 import org.ethereum.beacon.util.LRUCache;
 import org.javatuples.Pair;
 import tech.pegasys.artemis.ethereum.core.Hash32;
 import tech.pegasys.artemis.util.bytes.Bytes32;
-import tech.pegasys.artemis.util.bytes.BytesValue;
 import tech.pegasys.artemis.util.uint.UInt64;
 
-public class CachingSpecHelpers extends SpecHelpers {
-
-  public static SpecHelpers createWithSSZHasher(@Nonnull SpecConstants constants) {
-    Objects.requireNonNull(constants);
-
-    Function<BytesValue, Hash32> hashFunction = Hashes::keccak256;
-    ObjectHasher<Hash32> sszHasher = SSZObjectHasher.create(hashFunction);
-    return new SpecHelpers(constants, hashFunction, sszHasher);
-  }
+public class CachingBeaconChainSpec extends DelegatingBeaconChainSpec {
```

Comment:
Delegating wouldn't work for polymorphic invocations. E.g. `CachingSpecHelpers.get_permuted_list` would never be called

Reply:
That's true. Once call reaches delegate it never goes back to a class originated this call. Let me think around a little bit.